<?xml version="1.0"?>
<launch>
    <!-- 
        Launch file for ur10e_vs_node with UR driver and MoveIt integration
        功能：启动UR10e机器人驱动、MoveIt配置和视觉伺服节点
    -->

    <!-- Launch robot -->
    <!-- Use group tag to specify the namespace of the robot to avoid the  occupation of the root topic /joint_states -->
    <group ns="ur10e_robot">
        <!-- Set the robot IP address and calibration file-->
        <arg name="robot_ip" default="169.254.138.15" doc="IP address by which the robot can be reached."/>
        <arg name="kinematics_config" default="/home/vispci/my_robot_calibration.yaml" doc="Kinematics config file used for calibration correction. This will be used to verify the robot's calibration is matching the robot_description."/>
        <include file="$(find ur_robot_driver)/launch/ur10e_bringup.launch">
        <arg name="robot_ip" value="$(arg robot_ip)"/>
        <arg name="kinematics_config" value="$(arg kinematics_config)"/>
        </include>
    </group>

    <!-- Remap controller service for moveit to control the robot in the namespace "ur_robot" (I don't think this is an elegant implementation, and may update it later) -->
    <remap from="/controller_manager/list_controllers" to="/ur10e_robot/controller_manager/list_controllers"/>
    <remap from="/pos_joint_traj_controller/follow_joint_trajectory" to="/ur10e_robot/pos_joint_traj_controller/follow_joint_trajectory"/>

    <!-- Launch MoveIt -->
    <arg name="load_robot_description" default="true"/>
    <!-- By default, we are not in debug mode -->
    <arg name="debug" default="false" />
    <arg name="moveit_controller_manager" value="ros_control" />
    <!-- Set execution mode for fake execution controllers -->
    <arg name="fake_execution_type" default="interpolate" />
    <arg name="pipeline" default="ompl" />
    <!-- Load the move_group node -->
    <include file="$(find ur10e_moveit_config)/launch/demo.launch">
    <arg name="load_robot_description" value="$(arg load_robot_description)"/>
    </include>

    <!-- 启动视觉伺服节点 -->
    <node 
        name="ur10e_vs_node" 
        pkg="ur10e_vs_node" 
        type="ur10e_vs_node" 
        output="screen"  
        respawn="true"
        respawn_delay="5"
    >
        <!-- 视觉伺服参数 -->
        <param name="lambda" value="0.8" />
        <param name="max_linear_vel" value="0.1" />
        <param name="max_angular_vel" value="0.5" />
        <param name="aruco_marker_size" value="0.05" />
        <param name="display_enabled" value="true" />
    </node>

</launch>